### ISSUE TITLE
Extension vs composition for typed Scala Testkit

### ISSUE THREAD

**Issue by johanandren** [akka/akka#24598]
The `ActorTestKit` for typed was made a trait for mixin, mostly because that's how the old testkit is mostly used. But maybe we should rather encourage the same style as the typed Java API where the testkit is always a field inside your test.

Makes somewhat more OO-sense at least to me, that a testkit is something you have and can use, than that a test _is_ a testkit. But on the other hand the mixing is quite convenient.

We could also provide both, but I don't think we should have two ways to do the same unless there are clear situations where either one of them is a bad fit. Having a two ways that works pretty much for all use cases is just wasting the cognitive energy of our users.

**Comment by patriknw** [comment]
I had a few more arguments for composition:
* that style is already in several of the testing tools, such as TestProbe, BehaviorTestkit
* no risk of name clashes when we add more methods
* you can more easily create multiple in same test

**Comment by ktoso** [comment]
I guess the lifecycle thing causes the number of artifacts to increase much. I wonder if people would like it though;

**Comment by ktoso** [comment]
Having bumped into it a bit now, I feel this is very not nice:

```
extends ActorTestKit with TypedAkkaSpecWithShutdown 
```

so it starts a system, but does not manage its lifecycle properly -- this is bad. If we make it like this, it should also terminate the system (know about hooks in testing frameworks), then it'd be very good -- and accessible to new people.

**Comment by johanandren** [comment]
The example in the docs is to make an abstract base class for tests that does the shutdown, so that would be the `extends AkkaActorTestKitSpec` perhaps. And then name the mixin without the testkit something else.

**Comment by patriknw** [comment]
I agree that it's better that the lifecycle for create/shutdown is symmetrical, and that is actually another argument for why composition is a better. See the java example: https://doc.akka.io/docs/akka/current/typed/testing.html?language=java#asynchronous-testing

Regarding discoverability I think it's easier to discover things with `testkit.<ctrl-space>` than a global `<ctrl-space>`. 

Regarding test framework integration I think we can do it for scalatest and then use extension, since that is scalatest style anyway. The dependency would be provided. We should not support more frameworks, because we know how difficult it can be to get the dependency quickly for new scala versions.

**Comment by ktoso** [comment]
As long as we do at least a scalatest integration for lifecycle I'm fine with that -- my prime concern is to not have too much boilerplate when explaining it to new people

**Comment by ktoso** [comment]
(and not leaking systems of course)

**Comment by johanandren** [comment]
I'm convinced we should go for composition only.

**Comment by patriknw** [comment]
Conclusion: We agree that composition is the way and we will provide a Scala Test specific trait to glue ActorSystem lifecycle in public testkit, which we will also use internally.
